rpy2 conversion warnings

### Please make sure these conditions are met

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the latest version of pertpy.
- [ ] (optional) I have confirmed this bug exists on the main branch.

### Report

```pytb
 FAILED tests/tools/test_milo.py::test_da_nhoods_pvalues - DeprecationWarning: 
The activate and deactivate are deprecated. To set a conversion
context check the docstring for rpy2.robjects.conversion.Converter.context.
FAILED tests/tools/test_milo.py::test_da_nhoods_fdr - DeprecationWarning: 
The activate and deactivate are deprecated. To set a conversion
context check the docstring for rpy2.robjects.conversion.Converter.context.
FAILED tests/tools/test_milo.py::test_da_nhoods_default_contrast - DeprecationWarning: 
The activate and deactivate are deprecated. To set a conversion
context check the docstring for rpy2.robjects.conversion.Converter.context.
```


### Versions

```markdown

```

--- a/pertpy/tools/_milo.py
+++ b/pertpy/tools/_milo.py
@@ -379,7 +379,7 @@ class Milo:
             for var in variables:
                 if var in design_df.columns:
                     if pd.api.types.is_object_dtype(design_df[var]) or not pd.api.types.is_numeric_dtype(design_df[var]):
-                        design_df[var] = design_df[var].astype("category")
+                        design_df.loc[:, var] = design_df[var].astype("category")
 
             with localconverter(ro.default_converter + pandas2ri.converter):
                 design_r = pandas2ri.py2rpy(design_df)

--- a/pertpy/tools/_milo.py
+++ b/pertpy/tools/_milo.py
@@ -12,6 +12,8 @@ import seaborn as sns
 from anndata import AnnData
 from lamin_utils import logger
 from mudata import MuData
+import patsy
+
 
 from pertpy._doc import _doc_params, doc_common_plot_args
 
@@ -373,7 +375,6 @@ class Milo:
             if not add_intercept or model_contrasts is not None:
                 design = design + " + 0"
 
-            import patsy
             desc = patsy.ModelDesc.from_formula(design)
             variables = [term.name() for term in desc.rhs_termlist]
             for var in variables:

--- a/pertpy/tools/_milo.py
+++ b/pertpy/tools/_milo.py
@@ -7,7 +7,6 @@ from typing import TYPE_CHECKING, Literal
 import matplotlib.pyplot as plt
 import numpy as np
 import pandas as pd
-import patsy
 import scanpy as sc
 import seaborn as sns
 from anndata import AnnData
@@ -373,14 +372,7 @@ class Milo:
             # Define model matrix
             if not add_intercept or model_contrasts is not None:
                 design = design + " + 0"
-
-            desc = patsy.ModelDesc.from_formula(design)
-            variables = [term.name() for term in desc.rhs_termlist]
-            for var in variables:
-                if var in design_df.columns and (pd.api.types.is_object_dtype(design_df[var]) or not pd.api.types.is_numeric_dtype(design_df[var])):
-                    design_df.loc[:, var] = design_df[var].astype("category")
-
-
+            design_df = design_df.astype({col: "category" for col in design_df.select_dtypes(exclude=['number']).columns})  
             with localconverter(ro.default_converter + pandas2ri.converter):
                 design_r = pandas2ri.py2rpy(design_df)
             formula_r = stats.formula(design)

--- a/pertpy/tools/_milo.py
+++ b/pertpy/tools/_milo.py
@@ -12,10 +12,6 @@ import seaborn as sns
 from anndata import AnnData
 from lamin_utils import logger
 from mudata import MuData
-from rpy2.robjects import numpy2ri, pandas2ri
-from rpy2.robjects.conversion import localconverter
-import rpy2.robjects as ro
-from rpy2.robjects.vectors import FloatVector
 
 
 from pertpy._doc import _doc_params, doc_common_plot_args
@@ -369,6 +365,10 @@ class Milo:
             # Set up rpy2 to run edgeR
             edgeR, limma, stats, base = self._setup_rpy2()
 
+            from rpy2.robjects import numpy2ri, pandas2ri
+            from rpy2.robjects.conversion import localconverter
+            import rpy2.robjects as ro
+            from rpy2.robjects.vectors import FloatVector
             # Define model matrix
             if not add_intercept or model_contrasts is not None:
                 design = design + " + 0"
@@ -388,10 +388,8 @@ class Milo:
             dge = edgeR.estimateDisp(dge, model)
             fit = edgeR.glmQLFit(dge, model, robust=True)
             # Test
-            print("before n_coef")
             model_np = np.array(model)
             n_coef = model_np.shape[1]
-            print("after n_coef")
             if model_contrasts is not None:
                 r_str = """
                 get_model_cols <- function(design_df, design){
@@ -404,46 +402,31 @@ class Milo:
                 get_model_cols = STAP(r_str, "get_model_cols")
                 with localconverter(ro.default_converter + numpy2ri.converter + pandas2ri.converter):
                     model_mat_cols = get_model_cols.get_model_cols(design_df, design)
-                print("#412")
                 with localconverter(ro.default_converter + pandas2ri.converter + numpy2ri.converter):
                     model_df = pandas2ri.rpy2py(model)
                 model_df = pd.DataFrame(model_df)
-                print("#415")
                 model_df.columns = model_mat_cols
-                print("434")
                 try:
-                    print("437")
                     with localconverter(ro.default_converter + pandas2ri.converter):
                         mod_contrast = limma.makeContrasts(contrasts=model_contrasts, levels=model_df)
-                    print("440")
                 except ValueError:
                     logger.error("Model contrasts must be in the form 'A-B' or 'A+B'")
                     raise
-                print("#421: mod_contrast:", type(mod_contrast))
                 with localconverter(ro.default_converter + pandas2ri.converter + numpy2ri.converter):
                     res = base.as_data_frame(
                         edgeR.topTags(edgeR.glmQLFTest(fit, contrast=mod_contrast), sort_by="none", n=np.inf)
                     )
-                    print("#426: mod_contrast:", type(mod_contrast))
             else:
                 with localconverter(ro.default_converter + numpy2ri.converter + pandas2ri.converter):
                     res = base.as_data_frame(edgeR.topTags(edgeR.glmQLFTest(fit, coef=n_coef), sort_by="none", n=np.inf))
-                    print("#432: n_coef:", type(n_coef))
-            print("RES:", type(res))
             if not isinstance(res, pd.DataFrame):
-                print("before res")
                 res = pd.DataFrame(res)
-            print("after res")
-            print("Columns in res:", res.columns.tolist())
             res.columns = [col.replace("table.", "") for col in res.columns]
         # Save outputs
-        print("#469")
         res.index = sample_adata.var_names[keep_nhoods]  # type: ignore
-        print("#470")
         if any(col in sample_adata.var.columns for col in res.columns):
             sample_adata.var = sample_adata.var.drop(res.columns, axis=1)
         sample_adata.var = pd.concat([sample_adata.var, res], axis=1)
-        print("#474")
         # Run Graph spatial FDR correction
         self._graph_spatial_fdr(sample_adata, neighbors_key=adata.uns["nhood_neighbors_key"])
 

--- a/pertpy/tools/_milo.py
+++ b/pertpy/tools/_milo.py
@@ -372,7 +372,15 @@ class Milo:
             # Define model matrix
             if not add_intercept or model_contrasts is not None:
                 design = design + " + 0"
-            design_df["condition"] = design_df["condition"].astype("category")
+
+            import patsy
+            desc = patsy.ModelDesc.from_formula(design)
+            variables = [term.name() for term in desc.rhs_termlist]
+            for var in variables:
+                if var in design_df.columns:
+                    if pd.api.types.is_object_dtype(design_df[var]) or not pd.api.types.is_numeric_dtype(design_df[var]):
+                        design_df[var] = design_df[var].astype("category")
+
             with localconverter(ro.default_converter + pandas2ri.converter):
                 design_r = pandas2ri.py2rpy(design_df)
             formula_r = stats.formula(design)

--- a/pertpy/tools/_milo.py
+++ b/pertpy/tools/_milo.py
@@ -378,9 +378,9 @@ class Milo:
             desc = patsy.ModelDesc.from_formula(design)
             variables = [term.name() for term in desc.rhs_termlist]
             for var in variables:
-                if var in design_df.columns:
-                    if pd.api.types.is_object_dtype(design_df[var]) or not pd.api.types.is_numeric_dtype(design_df[var]):
-                        design_df.loc[:, var] = design_df[var].astype("category")
+                if var in design_df.columns and (pd.api.types.is_object_dtype(design_df[var]) or not pd.api.types.is_numeric_dtype(design_df[var])):
+                    design_df.loc[:, var] = design_df[var].astype("category")
+
 
             with localconverter(ro.default_converter + pandas2ri.converter):
                 design_r = pandas2ri.py2rpy(design_df)

--- a/pertpy/tools/_milo.py
+++ b/pertpy/tools/_milo.py
@@ -7,7 +7,6 @@ from typing import TYPE_CHECKING, Literal
 import matplotlib.pyplot as plt
 import numpy as np
 import pandas as pd
-import rpy2.robjects as ro
 import scanpy as sc
 import seaborn as sns
 from anndata import AnnData

--- a/pertpy/tools/_milo.py
+++ b/pertpy/tools/_milo.py
@@ -372,15 +372,13 @@ class Milo:
             # Define model matrix
             if not add_intercept or model_contrasts is not None:
                 design = design + " + 0"
-            # model = stats.model_matrix(object=stats.formula(design), data=design_df) # TODO desing_df is pandas, model is a NumPy array, SHOULD I convert it with model_np = np.array(model)?
-            design_df["condition"] = design_df["condition"].astype("category") # TODO is it safe to do this? R can't deal with "object" dtype
+            design_df["condition"] = design_df["condition"].astype("category")
             with localconverter(ro.default_converter + pandas2ri.converter):
                 design_r = pandas2ri.py2rpy(design_df)     
             formula_r = stats.formula(design)
             model = stats.model_matrix(object=formula_r, data=design_r)
 
             # Fit NB-GLM
-            # dge = edgeR.DGEList(counts=count_mat[keep_nhoods, :][:, keep_smp], lib_size=lib_size[keep_smp]) # TODO counts and lib_size are numpy, dge an R object
             counts_filtered = count_mat[np.ix_(keep_nhoods, keep_smp)]
             lib_size_filtered = lib_size[keep_smp]
             count_mat_r = numpy2ri.py2rpy(counts_filtered)
@@ -389,7 +387,6 @@ class Milo:
             dge = edgeR.calcNormFactors(dge, method="TMM")
             dge = edgeR.estimateDisp(dge, model)
             fit = edgeR.glmQLFit(dge, model, robust=True)
-            # TODO upto here we need model to be an R object
             # Test
             print("before n_coef")
             model_np = np.array(model)
@@ -405,18 +402,16 @@ class Milo:
                 from rpy2.robjects.packages import STAP
 
                 get_model_cols = STAP(r_str, "get_model_cols")
-                # model_mat_cols = get_model_cols.get_model_cols(design_df, design) # TODO design_df is pandas
                 with localconverter(ro.default_converter + numpy2ri.converter + pandas2ri.converter):
-                    model_mat_cols = get_model_cols.get_model_cols(design_df, design) # TODO model_mat_cols is now NumPy array
-                # model_df = pd.DataFrame(model) # TODO
+                    model_mat_cols = get_model_cols.get_model_cols(design_df, design)
                 print("#412")
                 with localconverter(ro.default_converter + pandas2ri.converter + numpy2ri.converter):
                     model_df = pandas2ri.rpy2py(model)
+                model_df = pd.DataFrame(model_df)
                 print("#415")
                 model_df.columns = model_mat_cols
                 print("434")
                 try:
-                    # mod_contrast = limma.makeContrasts(contrasts=model_contrasts, levels=model_df) # TODO model_df should be pandas 
                     print("437")
                     with localconverter(ro.default_converter + pandas2ri.converter):
                         mod_contrast = limma.makeContrasts(contrasts=model_contrasts, levels=model_df)
@@ -424,34 +419,23 @@ class Milo:
                 except ValueError:
                     logger.error("Model contrasts must be in the form 'A-B' or 'A+B'")
                     raise
-                # res = base.as_data_frame(
-                #     edgeR.topTags(edgeR.glmQLFTest(fit, contrast=mod_contrast), sort_by="none", n=np.inf)
-                # ) # TODO res is an R data frame object
                 print("#421: mod_contrast:", type(mod_contrast))
-                with localconverter(ro.default_converter + pandas2ri.converter):
+                with localconverter(ro.default_converter + pandas2ri.converter + numpy2ri.converter):
                     res = base.as_data_frame(
                         edgeR.topTags(edgeR.glmQLFTest(fit, contrast=mod_contrast), sort_by="none", n=np.inf)
                     )
                     print("#426: mod_contrast:", type(mod_contrast))
             else:
-                # res = base.as_data_frame(edgeR.topTags(edgeR.glmQLFTest(fit, coef=n_coef), sort_by="none", n=np.inf)) # TODO
-                print("#429: n_coef:", type(n_coef))
-                print("FIT:", type(fit))
-                print("N_COEF:", type(n_coef))
-                print("NP.INF:", type(np.inf))
                 with localconverter(ro.default_converter + numpy2ri.converter + pandas2ri.converter):
                     res = base.as_data_frame(edgeR.topTags(edgeR.glmQLFTest(fit, coef=n_coef), sort_by="none", n=np.inf))
                     print("#432: n_coef:", type(n_coef))
-            # from rpy2.robjects import conversion # TODO don't think i need this
-
-            # res = conversion.rpy2py(res) # TODO
             print("RES:", type(res))
             if not isinstance(res, pd.DataFrame):
                 print("before res")
                 res = pd.DataFrame(res)
             print("after res")
             print("Columns in res:", res.columns.tolist())
-            res.columns = [col.replace("table.", "") for col in res.columns] # TODO need to do this because the columns now start with table.
+            res.columns = [col.replace("table.", "") for col in res.columns]
         # Save outputs
         print("#469")
         res.index = sample_adata.var_names[keep_nhoods]  # type: ignore

--- a/pertpy/tools/_milo.py
+++ b/pertpy/tools/_milo.py
@@ -423,6 +423,7 @@ class Milo:
                     )
             if not isinstance(res, pd.DataFrame):
                 res = pd.DataFrame(res)
+            # The columns of res looks like e.g. table.A, table.B, so remove the prefix
             res.columns = [col.replace("table.", "") for col in res.columns]
         # Save outputs
         res.index = sample_adata.var_names[keep_nhoods]  # type: ignore

--- a/pertpy/tools/_milo.py
+++ b/pertpy/tools/_milo.py
@@ -12,6 +12,11 @@ import seaborn as sns
 from anndata import AnnData
 from lamin_utils import logger
 from mudata import MuData
+from rpy2.robjects import numpy2ri, pandas2ri
+from rpy2.robjects.conversion import localconverter
+import rpy2.robjects as ro
+from rpy2.robjects.vectors import FloatVector
+
 
 from pertpy._doc import _doc_params, doc_common_plot_args
 
@@ -367,16 +372,29 @@ class Milo:
             # Define model matrix
             if not add_intercept or model_contrasts is not None:
                 design = design + " + 0"
-            model = stats.model_matrix(object=stats.formula(design), data=design_df)
+            # model = stats.model_matrix(object=stats.formula(design), data=design_df) # TODO desing_df is pandas, model is a NumPy array, SHOULD I convert it with model_np = np.array(model)?
+            design_df["condition"] = design_df["condition"].astype("category") # TODO is it safe to do this? R can't deal with "object" dtype
+            with localconverter(ro.default_converter + pandas2ri.converter):
+                design_r = pandas2ri.py2rpy(design_df)     
+            formula_r = stats.formula(design)
+            model = stats.model_matrix(object=formula_r, data=design_r)
 
             # Fit NB-GLM
-            dge = edgeR.DGEList(counts=count_mat[keep_nhoods, :][:, keep_smp], lib_size=lib_size[keep_smp])
+            # dge = edgeR.DGEList(counts=count_mat[keep_nhoods, :][:, keep_smp], lib_size=lib_size[keep_smp]) # TODO counts and lib_size are numpy, dge an R object
+            counts_filtered = count_mat[np.ix_(keep_nhoods, keep_smp)]
+            lib_size_filtered = lib_size[keep_smp]
+            count_mat_r = numpy2ri.py2rpy(counts_filtered)
+            lib_size_r = FloatVector(lib_size_filtered)
+            dge = edgeR.DGEList(counts=count_mat_r, lib_size=lib_size_r)
             dge = edgeR.calcNormFactors(dge, method="TMM")
             dge = edgeR.estimateDisp(dge, model)
             fit = edgeR.glmQLFit(dge, model, robust=True)
-
+            # TODO upto here we need model to be an R object
             # Test
-            n_coef = model.shape[1]
+            print("before n_coef")
+            model_np = np.array(model)
+            n_coef = model_np.shape[1]
+            print("after n_coef")
             if model_contrasts is not None:
                 r_str = """
                 get_model_cols <- function(design_df, design){
@@ -387,32 +405,61 @@ class Milo:
                 from rpy2.robjects.packages import STAP
 
                 get_model_cols = STAP(r_str, "get_model_cols")
-                model_mat_cols = get_model_cols.get_model_cols(design_df, design)
-                model_df = pd.DataFrame(model)
+                # model_mat_cols = get_model_cols.get_model_cols(design_df, design) # TODO design_df is pandas
+                with localconverter(ro.default_converter + numpy2ri.converter + pandas2ri.converter):
+                    model_mat_cols = get_model_cols.get_model_cols(design_df, design) # TODO model_mat_cols is now NumPy array
+                # model_df = pd.DataFrame(model) # TODO
+                print("#412")
+                with localconverter(ro.default_converter + pandas2ri.converter + numpy2ri.converter):
+                    model_df = pandas2ri.rpy2py(model)
+                print("#415")
                 model_df.columns = model_mat_cols
+                print("434")
                 try:
-                    mod_contrast = limma.makeContrasts(contrasts=model_contrasts, levels=model_df)
+                    # mod_contrast = limma.makeContrasts(contrasts=model_contrasts, levels=model_df) # TODO model_df should be pandas 
+                    print("437")
+                    with localconverter(ro.default_converter + pandas2ri.converter):
+                        mod_contrast = limma.makeContrasts(contrasts=model_contrasts, levels=model_df)
+                    print("440")
                 except ValueError:
                     logger.error("Model contrasts must be in the form 'A-B' or 'A+B'")
                     raise
-                res = base.as_data_frame(
-                    edgeR.topTags(edgeR.glmQLFTest(fit, contrast=mod_contrast), sort_by="none", n=np.inf)
-                )
+                # res = base.as_data_frame(
+                #     edgeR.topTags(edgeR.glmQLFTest(fit, contrast=mod_contrast), sort_by="none", n=np.inf)
+                # ) # TODO res is an R data frame object
+                print("#421: mod_contrast:", type(mod_contrast))
+                with localconverter(ro.default_converter + pandas2ri.converter):
+                    res = base.as_data_frame(
+                        edgeR.topTags(edgeR.glmQLFTest(fit, contrast=mod_contrast), sort_by="none", n=np.inf)
+                    )
+                    print("#426: mod_contrast:", type(mod_contrast))
             else:
-                res = base.as_data_frame(edgeR.topTags(edgeR.glmQLFTest(fit, coef=n_coef), sort_by="none", n=np.inf))
+                # res = base.as_data_frame(edgeR.topTags(edgeR.glmQLFTest(fit, coef=n_coef), sort_by="none", n=np.inf)) # TODO
+                print("#429: n_coef:", type(n_coef))
+                print("FIT:", type(fit))
+                print("N_COEF:", type(n_coef))
+                print("NP.INF:", type(np.inf))
+                with localconverter(ro.default_converter + numpy2ri.converter + pandas2ri.converter):
+                    res = base.as_data_frame(edgeR.topTags(edgeR.glmQLFTest(fit, coef=n_coef), sort_by="none", n=np.inf))
+                    print("#432: n_coef:", type(n_coef))
+            # from rpy2.robjects import conversion # TODO don't think i need this
 
-            from rpy2.robjects import conversion
-
-            res = conversion.rpy2py(res)
+            # res = conversion.rpy2py(res) # TODO
+            print("RES:", type(res))
             if not isinstance(res, pd.DataFrame):
+                print("before res")
                 res = pd.DataFrame(res)
-
+            print("after res")
+            print("Columns in res:", res.columns.tolist())
+            res.columns = [col.replace("table.", "") for col in res.columns] # TODO need to do this because the columns now start with table.
         # Save outputs
+        print("#469")
         res.index = sample_adata.var_names[keep_nhoods]  # type: ignore
+        print("#470")
         if any(col in sample_adata.var.columns for col in res.columns):
             sample_adata.var = sample_adata.var.drop(res.columns, axis=1)
         sample_adata.var = pd.concat([sample_adata.var, res], axis=1)
-
+        print("#474")
         # Run Graph spatial FDR correction
         self._graph_spatial_fdr(sample_adata, neighbors_key=adata.uns["nhood_neighbors_key"])
 
@@ -657,11 +704,8 @@ class Milo:
         self,
     ):
         """Set up rpy2 to run edgeR."""
-        from rpy2.robjects import numpy2ri, pandas2ri
         from rpy2.robjects.packages import importr
 
-        numpy2ri.activate()
-        pandas2ri.activate()
         edgeR = self._try_import_bioc_library("edgeR")
         limma = self._try_import_bioc_library("limma")
         stats = importr("stats")

